File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_10.java
package org.twyszomirski.aoc.solutions;


import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.io.IO.println;

public class Day_10 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_10.txt"));
        println("======== Day 10 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var data = getData(lines);
        var startingPoints = data.stream().flatMap(Collection::stream).filter(it -> it.value == 0);

        var trails = startingPoints.map(start -> findTrails(data, Collections.singletonList(start)));
        var trailHeads = trails.map(t -> t.stream().map(List::getLast).collect(Collectors.toSet())).map(Collection::size);
        var sum = trailHeads.mapToInt(it -> it).sum();
        println(sum);
    }

    private static void part_2(List<String> lines) {
        var data = getData(lines);
        var startingPoints = data.stream().flatMap(Collection::stream).filter(it -> it.value == 0);

        var trails = startingPoints.map(start -> findTrails(data, Collections.singletonList(start)));
        var sum = trails.map(List::size).mapToInt(it -> it).sum();
        println(sum);
    }

    private static List<List<TrailPoint>> findTrails(List<List<TrailPoint>> grid, List<TrailPoint> head) {
        if (head.size() == 10) {
            return Collections.singletonList(head);
        }
        var result = new ArrayList<List<TrailPoint>>();

        var left = getNeighbour(grid, head.getLast(), -1, 0);
        var right = getNeighbour(grid, head.getLast(), 1, 0);
        var up = getNeighbour(grid, head.getLast(), 0, -1);
        var down = getNeighbour(grid, head.getLast(), 0, 1);

        Stream.of(left, right, up, down).filter(Objects::nonNull).forEach(
                point -> {
                    if (point.value - 1 == head.getLast().value) {
                        var copy = head.stream().collect(Collectors.toList());
                        copy.add(point);
                        result.addAll(findTrails(grid, copy));
                    }
                });
        return result;
    }

    private static TrailPoint getNeighbour(List<List<TrailPoint>> grid, TrailPoint point, int horizontal, int vertical) {
        var column = point.column + horizontal;
        var row = point.row + vertical;

        if (row >= 0 && column >= 0 && row < grid.size() && column < grid.getFirst().size()) {
            return grid.get(row).get(column);
        }

        return null;
    }


    private static List<List<TrailPoint>> getData(List<String> lines) {
        var grid = new ArrayList<List<TrailPoint>>();
        for (int row = 0; row < lines.size(); row++) {
            var currentLine = lines.get(row).split("");
            grid.add(new ArrayList<>());
            for (int column = 0; column < currentLine.length; column++) {
                grid.getLast().add(new TrailPoint(row, column, Objects.equals(currentLine[column], ".") ? -1 : Integer.parseInt(currentLine[column])));
            }
        }
        return grid;
    }

    record TrailPoint(Integer row, Integer column, Integer value) {
    }

}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_10.kt
package org.twyszomirski.aoc.solutions

import org.twyszomirski.aoc.solutions.Day_10.TrailPoint
import java.io.File


class Day_10 {

    fun solve() {
        val lines = File("src/main/resources/input_day_10.txt").readLines()
        println("======== Day 10 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val data = getData(input)
        val starters = data.flatten().filter { it.inclination == 0 }
        val sum = starters.map { listOf(it) }.map { getTrails(data, it) }
            .map { it.map { it.last() }.toSet() }.sumOf { it.size }

        println(sum)
    }


    fun part2(input: List<String>) {
        val data = getData(input)
        val starters = data.flatten().filter { it.inclination == 0 }
        val sum = starters.map { listOf(it) }.map { getTrails(data, it) }
            .sumOf { it.size }

        println(sum)
    }

    private fun getTrails(matrix: TrailMatrix, trail: List<TrailPoint>): List<Trail> {
        val head = trail.last()
        val candidates = listOf(
            getCandidate(matrix, head, 0, -1),
            getCandidate(matrix, head, 1, 0),
            getCandidate(matrix, head, 0, 1),
            getCandidate(matrix, head, -1, 0),
        ).filter { it != null }
            .filter { it!!.inclination == head.inclination + 1 }
            .map { trail + listOf(it!!) }
        if (trail.size == 9) {
            return candidates
        }
        return candidates.flatMap { it -> getTrails(matrix, it) }
    }

    fun getCandidate(matrix: TrailMatrix, previous: TrailPoint, horizontal: Int, vertical: Int): TrailPoint? {
        val nextRow = previous.row + vertical
        val nextCol = previous.column + horizontal
        if (nextRow >= 0 && nextRow < matrix.size && nextCol >= 0 && nextCol < matrix.first().size) {
            return matrix[nextRow][nextCol]
        }
        return null
    }

    private fun getData(lines: List<String>): TrailMatrix {
        return lines.mapIndexed { row, line ->
            line.split("").filter { it.isNotBlank() }
                .map { if (".".equals(it)) -1 else it.toInt() }
                .mapIndexed { col, p -> TrailPoint(row, col, p) }.toList()
        }.toList()
    }


    data class TrailPoint(val row: Int, val column: Int, val inclination: Int)

}

typealias TrailMatrix = List<List<TrailPoint>>
typealias Trail = List<TrailPoint>
